lib/js-yaml/dumper.js

```diff
@@ -682,7 +684,31 @@ function detectType(state, object, explicit) {
 // Serializes `object` and writes it to global `result`.
 // Returns true on success, or false on invalid object.
 //
-function writeNode(state, level, object, block, compact, iskey) {
+function writeNode(state, level, oobject, block, compact, iskey) {
+  // replacement, sjpt
+  /** test case:
+    obj = {x:3, y:7, zzz: {q:3, toJSON: function(m) {return this.q || 7}}};
+    repl = (k,v)=> v === 3 ? k + 993 : v
+    yyy = jsyaml.safeDump(obj, {replacer: repl , localReplacer: 'toJSON', skipInvalid: true})
+    jjj = JSON.stringify(obj, repl)
+    console.log('yyy => ', yyy); console.log('jjj => ', jjj)
+
+  failing test case, because keys get processed through this as well as values
+    obj = {x:3, y:7, zzz: {q:3, toJSON: function(m) {return this.q || 7}}};
+    repl = (k,v)=> v === 'y' ? k + 993 : v
+    yyy = jsyaml.safeDump(obj, {replacer: repl , localReplacer: 'toJSON', skipInvalid: true})
+    jjj = JSON.stringify(obj, repl)
+    console.log('yyy => ', yyy); console.log('jjj => ', jjj)
+  ***/
+  let object = oobject; // keep original for debug
+  if (typeof object === 'object' && object[state.localReplacer]) {
+    object = object[state.localReplacer].apply(object);
```

Comment:
This is not called  correctly:

> JSON.stringify() calls toJSON with one parameter:
>
> * if this object is a property value, the property name
> * if it is in an array, the index in the array, as a string
> * an empty string if JSON.stringify() was directly called on this object


 do it like this:

```js
object_ = object[state.localReplacer](key || '');
```

**NOTE:** we have a problem here, `writeNode` is not aware of the key it is writing.

Also, not each invocation of  `writeNode` writes a value, some invocations just write keys, see the `isKey` parameter. If `isKey` is true, we should never even consider calling any replacer.